contentcuration/contentcuration/view/node_views.py

```diff
@@ -0,0 +1,290 @@
+    File.objects.bulk_create(to_create["node_files"] + to_create["assessment_files"])
+
+    return new_node
+
+def _duplicate_node_bulk_recursive(node, sort_order, parent, channel_id, to_create, level=0):
+
+    if isinstance(node, int) or isinstance(node, basestring):
+        node = ContentNode.objects.get(pk=node)
+
+    if isinstance(parent, int) or isinstance(parent, basestring):
+        parent = ContentNode.objects.get(pk=parent)
+
+    # clone the model (in-memory) and update the fields on the cloned model
+    new_node = copy.copy(node)
+    new_node.id = None
+    new_node.tree_id = parent.tree_id
+    new_node.parent = parent
+    new_node.sort_order = sort_order or node.sort_order
+    new_node.changed = True
+    new_node.cloned_source = node
+    new_node.source_channel_id = node.get_channel().id if node.get_channel() else None
+    new_node.node_id = uuid.uuid4().hex
+    new_node.source_node_id = node.node_id
+
+    # There might be some legacy nodes that don't have these, so ensure they are added
+    if not new_node.original_channel_id or not new_node.original_source_node_id:
+        original_node = node.get_original_node()
+        original_channel = original_node.get_channel()
+        new_node.original_channel_id = original_channel.id if original_channel else None
+        new_node.original_source_node_id = original_node.node_id
+
+    # store the new unsaved model in a list, at the appropriate level, for later creation
+    while len(to_create["nodes"]) <= level:
+        to_create["nodes"].append([])
+    to_create["nodes"][level].append(new_node)
+
+    # find or create any tags that are needed, and store them under _meta on the node so we can add them to it later
+    new_node._meta.tags_to_add = []
+    for tag in node.tags.all():
+        new_tag, is_new = ContentTag.objects.get_or_create(
+            tag_name=tag.tag_name,
+            channel_id=channel_id,
+        )
+        new_node._meta.tags_to_add.append(new_tag)
+
+    # clone the file objects for later saving
+    for fobj in node.files.all():
+        f = duplicate_file(fobj, node=new_node, save=False)
+        to_create["node_files"].append(f)
+
+    # copy assessment item objects, and associated files
+    for aiobj in node.assessment_items.prefetch_related("files").all():
+        aiobj_copy = copy.copy(aiobj)
+        aiobj_copy.id = None
+        aiobj_copy.contentnode = new_node
+        to_create["assessments"].append(aiobj_copy)
+        for fobj in aiobj.files.all():
+            f = duplicate_file(fobj, assessment_item=aiobj_copy, save=False)
+            to_create["assessment_files"].append(f)
+
+    # recurse down the tree and clone the children
+    for child in node.children.all():
+        _duplicate_node_bulk_recursive(node=child, sort_order=None, parent=new_node, channel_id=channel_id, to_create=to_create, level=level+1)
+
+    return new_node
+
+def _duplicate_node(node, sort_order=None, parent=None, channel_id=None):
```

Comment:
Was this reintroduced here, or did I not delete it in my PR? When would we use this one?